src/hotp/hotp.c

```diff
@@ -493,11 +493,13 @@ uint8_t* page = (uint8_t *) current_slot_address;
 
     memcpy (page_buffer, page, SLOT_PAGE_SIZE);
 
-    // check if the secret from the tool is empty and if it is use the old
-    // secret
-uint8_t* secret = (uint8_t *) (data + SECRET_OFFSET);
-
-    if (secret[0] == 0)
+    // Check if the secret from the tool is empty and if it is use the old secret
+    // Secret could begin with 0x00, so checking the whole secret before keeping the old one in mandatory
+    // The while last loop points secret[-1]. Well, it's right, == result does not matter and mem exits, we need a -1
+    uint8_t* secret = (uint8_t *) (data + SECRET_OFFSET);
+    int8_t i = 20;
+    while (i >= 0 && secret[--i] == 0x00);
```

Comment:
Stupid question: Why does the App ever send an empty secret?

While I agree that 20*0x00 is a very practical fix, I'm not a fan of accessing
memory that doesn't belong to us for a variety of reasons.
I would also prefer something less subtle. How about something like this:
```
// Check if the secret from the tool is empty.
uint8_t empty = 1;
for (int i = 0; empty && i<20; i++) {
  if (secret[i]!=0x00) {
    empty = 0;
  }
}
// If the secret from the tool is empty keep the old one.
if (!empty) {
```